File: Ch6/02.play_again/play_again0.c
/*
  play_again0.c
  purpose: ask if user want another transaction
  method: ask a question, wait for yes/no answer
  returns: 0=>yes, 1=>no
  better: eliminate need to press return
*/

#include <stdio.h>

#define QUESTION "Do you want another transaction"

/*
  purpose: ask a question and wait for a y/n answer
  method: use getchar and ignore non y/n answers
  returns: 0=>yes, 1=>no
*/
int get_response(char * question) {
    printf("%s (y/n)", QUESTION);
    while(1) {
        switch (getchar() ) {
            case 'y':
            case 'Y':
                return 0;
            case 'n':
            case 'N':
            case EOF:
                return 1;
        }
    }
}

int main() {
    int response = get_response(QUESTION);
    return response;
}

File: Ch6/02.play_again/play_again0.rs
use std::process;
use std::io::{self, stdin, Read, Write};

#[derive(PartialEq)]
enum Answer{
    YES,
    NO,
}

fn get_response() -> Answer {
    print!("Do you want another transaction(y/n)");
    io::stdout().flush().unwrap();

    for n in stdin().bytes().map(|x|x.expect("Can not read from stdin")) {
        if n == 'y' as u32 as u8 || n == 'Y' as u32 as u8 {
            return Answer::YES;
        }else if n=='n' as u32 as u8 || n=='N' as u32 as u8 {
            return Answer::NO;
        }else{
            continue;
        }
    }
    unreachable!()
}


fn main() {
    if get_response() == Answer::NO {
        process::exit(1);
    }
}